File: services/application/models/Setting_model.php
<?php
defined('BASEPATH') or exit('No direct script access allowed');

class Setting_model extends CI_Model
{

    public function __construct()
    {
        parent::__construct();
        $this->load->database();
        $this->load->library(['ion_auth', 'form_validation']);
        $this->load->helper(['url', 'language', 'function_helper', 'timezone_helper']);
    }

    public function update_system_setting($post)
    {
        $post = escape_array($post);

        $system_data = [
            'system_configurations' => $post['system_configurations'],
            'system_timezone_gmt' => $post['system_timezone_gmt'],
            'system_configurations_id' => $post['system_configurations_id'],
            'app_name' => $post['app_name'],
            'support_number' => $post['support_number'],
            'support_email' => $post['support_email'],
            'current_version' => $post['current_version'],
            'current_version_ios' => $post['current_version_ios'],
            'is_version_system_on' => (isset($post['is_version_system_on'])) ? '1' : '0',
            'otp_login' => (isset($post['otp_login'])) ? '1' : '0',
            'google_login' => (isset($post['google_login'])) ? '1' : '0',
            // 'facebook_login' => (isset($post['facebook_login'])) ? '1' : '0',
            'apple_login' => (isset($post['apple_login'])) ? '1' : '0',
            'currency' => $post['currency'],
            'system_timezone' => $post['system_timezone'],
            'is_refer_earn_on' => (isset($post['is_refer_earn_on'])) ? '1' : '0',
            'tax' => (isset($post['tax'])) ? $post['tax'] : '0',
            'is_email_setting_on' => (isset($post['is_email_setting_on'])) ? '1' : '0',
            'google_map_api_key' => $post['google_map_api_key'],
            'google_map_javascript_api_key' => $post['google_map_javascript_api_key'],
            'min_refer_earn_order_amount' => $post['min_refer_earn_order_amount'],
            'refer_earn_bonus' => $post['refer_earn_bonus'],
            'refer_earn_method' => $post['refer_earn_method'],
            'max_refer_earn_amount' => $post['max_refer_earn_amount'],
            'refer_earn_bonus_times' => $post['refer_earn_bonus_times'],
            'minimum_cart_amt' => $post['minimum_cart_amt'],
            'low_stock_limit' => (isset($post['low_stock_limit'])) ? $post['low_stock_limit'] : '5',
            'max_items_cart' => $post['max_items_cart'],
            'is_rider_otp_setting_on' => (isset($post['is_rider_otp_setting_on'])) ? '1' : '0',
            'is_app_maintenance_mode_on' => (isset($post['is_app_maintenance_mode_on'])) ? '1' : '0',
            'is_rider_app_maintenance_mode_on' => (isset($post['is_rider_app_maintenance_mode_on'])) ? '1' : '0',
            // 'is_partner_app_maintenance_mode_on' => (isset($post['is_partner_app_maintenance_mode_on'])) ? '1' : '0',
            // 'is_web_maintenance_mode_on' => (isset($post['is_web_maintenance_mode_on'])) ? '1' : '0',
            'supported_locals' => (isset($post['supported_locals'])) ?  $post['supported_locals'] : '',
        ];

        $main_image_name = $post['logo'];
        $rider_main_image_name = $post['rider_logo'];

        $favicon_image_name = $post['favicon'];
        $rider_favicon_image_name = $post['rider_favicon'];
        $cover_image_name = $post['cover_image'];
        $rider_cover_image_name = $post['rider_cover_image'];
        // $light_image_name = $post['light_logo'];

        $system_data = json_encode($system_data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'system_settings'
        ));
        $count = $query->num_rows();
        if ($main_image_name != NULL && !empty($main_image_name)) {
            $logo_res = $this->db->get_where('settings', array(
                'variable' => 'logo'
            ));
            $logo_count = $logo_res->num_rows();
            if ($logo_count == 0) {
                $this->db->insert('settings', ['value' => $main_image_name, 'variable' => 'logo']);
            } else {
                $this->db->set('value', $main_image_name)->where('variable', 'logo')->update('settings');
            }
        }
        // rider logo
        if ($rider_main_image_name != NULL && !empty($rider_main_image_name)) {
            $rider_logo_res = $this->db->get_where('settings', array(
                'variable' => 'rider_logo'
            ));
            $rider_logo_count = $rider_logo_res->num_rows();
            if ($rider_logo_count == 0) {
                $this->db->insert('settings', ['value' => $rider_main_image_name, 'variable' => 'rider_logo']);
            } else {
                $this->db->set('value', $rider_main_image_name)->where('variable', 'rider_logo')->update('settings');
            }
        }
        // end

        // if ($light_image_name != NULL && !empty($light_image_name)) {
        //     $logo_res = $this->db->get_where('settings', array(
        //         'variable' => 'light_logo'
        //     ));
        //     $logo_count = $logo_res->num_rows();
        //     if ($logo_count == 0) {
        //         $this->db->insert('settings', ['value' => $light_image_name, 'variable' => 'light_logo']);
        //     } else {
        //         $this->db->set('value', $light_image_name)->where('variable', 'light_logo')->update('settings');
        //     }
        // }
        if ($favicon_image_name != NULL && !empty($favicon_image_name)) {
            $favicon_res = $this->db->get_where('settings', array(
                'variable' => 'favicon'
            ));
            $favicon_count = $favicon_res->num_rows();
            if ($favicon_count == 0) {
                $this->db->insert('settings', ['value' => $favicon_image_name, 'variable' => 'favicon']);
            } else {
                $this->db->set('value', $favicon_image_name)->where('variable', 'favicon')->update('settings');
            }
        }

        // rider fevicon
        if ($rider_favicon_image_name != NULL && !empty($rider_favicon_image_name)) {
            $rider_favicon_res = $this->db->get_where('settings', array(
                'variable' => 'rider_favicon'
            ));
            $rider_favicon_count = $rider_favicon_res->num_rows();
            if ($rider_favicon_count == 0) {
                $this->db->insert('settings', ['value' => $rider_favicon_image_name, 'variable' => 'rider_favicon']);
            } else {
                $this->db->set('value', $rider_favicon_image_name)->where('variable', 'rider_favicon')->update('settings');
            }
        }
        // end

        if ($cover_image_name != NULL && !empty($cover_image_name)) {
            // print_r("hello");
            $cover_image_res = $this->db->get_where('settings', array(
                'variable' => 'cover_image'
            ));
            $cover_image_count = $cover_image_res->num_rows();
            if ($cover_image_count == 0) {
                $this->db->insert('settings', ['value' => $cover_image_name, 'variable' => 'cover_image']);
            } else {
                $this->db->set('value', $cover_image_name)->where('variable', 'cover_image')->update('settings');
            }
        }

        // rider cover image
        if ($rider_cover_image_name != NULL && !empty($rider_cover_image_name)) {

            $rider_cover_image_res = $this->db->get_where('settings', array(
                'variable' => 'rider_cover_image'
            ));
            $rider_cover_image_count = $rider_cover_image_res->num_rows();
            if ($rider_cover_image_count == 0) {
                $this->db->insert('settings', ['value' => $rider_cover_image_name, 'variable' => 'rider_cover_image']);
            } else {
                $this->db->set('value', $rider_cover_image_name)->where('variable', 'rider_cover_image')->update('settings');
            }
        }

        // end

        if ($count === 0) {
            $data = array(
                'variable' => 'system_settings',
                'value' => $system_data
            );
            $this->db->insert('settings', $data);
            $this->db->insert('settings', ['value' => $post['currency']]);
        } else {
            $this->db->set('value', $system_data)->where('variable', 'system_settings')->update('settings');
            $this->db->set('value', $post['currency'])->where('variable', 'currency')->update('settings');
        }
    }


    public function update_payment_method($post)
    {

        $post = escape_array($post);

        $payment_data = array();
        $payment_data['paypal_payment_method'] = isset($post['paypal_payment_method']) ? '1' : '0';
        $payment_data['paypal_mode'] = isset($post['paypal_mode']) && !empty($post['paypal_mode']) ? $post['paypal_mode'] : '';
        $payment_data['paypal_business_email'] = isset($post['paypal_business_email']) && !empty($post['paypal_business_email']) ? $post['paypal_business_email'] : '';
        $payment_data['currency_code'] = isset($post['currency_code']) && !empty($post['currency_code']) ? $post['currency_code'] : '';

        $payment_data['razorpay_payment_method'] = isset($post['razorpay_payment_method']) ? '1' : '0';
        $payment_data['razorpay_key_id'] = isset($post['razorpay_key_id']) && !empty($post['razorpay_key_id']) ? $post['razorpay_key_id'] : '';
        $payment_data['razorpay_secret_key'] = isset($post['razorpay_secret_key']) && !empty($post['razorpay_secret_key']) ? $post['razorpay_secret_key'] : '';
        $payment_data['refund_webhook_secret_key'] = isset($post['refund_webhook_secret_key']) && !empty($post['refund_webhook_secret_key']) ? $post['refund_webhook_secret_key'] : '';


        $payment_data['paystack_payment_method'] = isset($post['paystack_payment_method']) ? '1' : '0';
        $payment_data['paystack_key_id'] = isset($post['paystack_key_id']) && !empty($post['paystack_key_id']) ? $post['paystack_key_id'] : '';
        $payment_data['paystack_secret_key'] = isset($post['paystack_secret_key']) && !empty($post['paystack_secret_key']) ? $post['paystack_secret_key'] : '';


        $payment_data['stripe_payment_method'] = isset($post['stripe_payment_method']) ? '1' : '0';
        $payment_data['stripe_payment_mode'] = isset($post['stripe_payment_mode']) ? $post['stripe_payment_mode'] : 'test';
        $payment_data['stripe_publishable_key'] = isset($post['stripe_publishable_key']) && !empty($post['stripe_publishable_key']) ? $post['stripe_publishable_key'] : '';
        $payment_data['stripe_secret_key'] = isset($post['stripe_secret_key']) && !empty($post['stripe_secret_key']) ? $post['stripe_secret_key'] : '';
        $payment_data['stripe_webhook_secret_key'] = isset($post['stripe_webhook_secret_key']) && !empty($post['stripe_webhook_secret_key']) ? $post['stripe_webhook_secret_key'] : '';
        $payment_data['stripe_currency_code'] = isset($post['stripe_currency_code']) && !empty($post['stripe_currency_code']) ? $post['stripe_currency_code'] : '';

        $payment_data['flutterwave_payment_method'] = isset($post['flutterwave_payment_method']) ? '1' : '0';
        $payment_data['flutterwave_public_key'] = isset($post['flutterwave_public_key']) && !empty($post['flutterwave_public_key']) ? $post['flutterwave_public_key'] : '';
        $payment_data['flutterwave_secret_key'] = isset($post['flutterwave_secret_key']) && !empty($post['flutterwave_secret_key']) ? $post['flutterwave_secret_key'] : '';
        $payment_data['flutterwave_encryption_key'] = isset($post['flutterwave_encryption_key']) && !empty($post['flutterwave_encryption_key']) ? $post['flutterwave_encryption_key'] : '';
        $payment_data['flutterwave_currency_code'] = isset($post['flutterwave_currency_code']) && !empty($post['flutterwave_currency_code']) ? $post['flutterwave_currency_code'] : '';

        $payment_data['paytm_payment_method'] = isset($post['paytm_payment_method']) ? '1' : '0';
        $payment_data['paytm_payment_mode'] = isset($post['paytm_payment_mode']) && !empty($post['paytm_payment_mode']) ? $post['paytm_payment_mode'] : '';
        $payment_data['paytm_merchant_key'] = isset($post['paytm_merchant_key']) && !empty($post['paytm_merchant_key']) ? $post['paytm_merchant_key'] : '';
        $payment_data['paytm_merchant_id'] = isset($post['paytm_merchant_id']) && !empty($post['paytm_merchant_id']) ? $post['paytm_merchant_id'] : '';
        $payment_data['paytm_website'] = isset($post['paytm_payment_mode']) && $post['paytm_payment_mode'] == 'production' ? $post['paytm_website'] : 'WEBSTAGING';
        $payment_data['paytm_industry_type_id'] = isset($post['paytm_payment_mode']) && $post['paytm_payment_mode'] == 'production' ? $post['paytm_industry_type_id'] : 'Retail';

        $payment_data['midtrans_payment_mode'] = isset($post['midtrans_payment_mode']) && !empty($post['midtrans_payment_mode']) ? $post['midtrans_payment_mode'] : '';
        $payment_data['midtrans_payment_method'] = isset($post['midtrans_payment_method']) ? '1' : '0';
        $payment_data['midtrans_client_key'] = isset($post['midtrans_client_key']) && !empty($post['midtrans_client_key']) ? $post['midtrans_client_key'] : '';
        $payment_data['midtrans_merchant_id'] = isset($post['midtrans_merchant_id']) && !empty($post['midtrans_merchant_id']) ? $post['midtrans_merchant_id'] : '';
        $payment_data['midtrans_server_key'] = isset($post['midtrans_server_key']) && !empty($post['midtrans_server_key']) ? $post['midtrans_server_key'] : '';

        $payment_data['cod_method'] = isset($post['cod_method']) ? '1' : '0';

        $payment_data = json_encode($payment_data);

        $query = $this->db->get_where('settings', array(
            'variable' => 'payment_method'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'payment_method',
                'value' => $payment_data
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $payment_data)->where('variable', 'payment_method')->update('settings');
        }
    }

    public function update_fcm_details($post)
    {
        $post = escape_array($post);

        $query = $this->db->get_where('settings', array(
            'variable' => 'fcm_server_key'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'fcm_server_key',
                'value' => $post['fcm_server_key']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $post['fcm_server_key'])->where('variable', 'fcm_server_key')->update('settings');
        }
    }

    public function update_contact_details($post)
    {
        $post = escape_array($post);

        $query = $this->db->get_where('settings', array(
            'variable' => 'contact_us'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'contact_us',
                'value' => $post['contact_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $post['contact_input_description'])->where('variable', 'contact_us')->update('settings');
        }
    }

    public function update_privacy_policy($post)
    {
        $post = escape_array($post);

        $query = $this->db->get_where('settings', array(
            'variable' => 'privacy_policy'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'privacy_policy',
                'value' => $post['privacy_policy_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $post['privacy_policy_input_description'])->where('variable', 'privacy_policy')->update('settings');
        }
    }

    public function update_terms_n_condtions($post)
    {
        $post = escape_array($post);

        $query = $this->db->get_where('settings', array(
            'variable' => 'terms_conditions'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'terms_conditions',
                'value' => $post['terms_n_conditions_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $post['terms_n_conditions_input_description'])->where('variable', 'terms_conditions')->update('settings');
        }
    }

    public function update_about_us($post)
    {
        $query = $this->db->get_where('settings', array(
            'variable' => 'about_us'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'about_us',
                'value' => $post['about_us_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $post['about_us_input_description'])->where('variable', 'about_us')->update('settings');
        }
    }

    public function update_email_settings($data)
    {
        $data = escape_array($data);
        $email_data = json_encode($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'email_settings'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'email_settings',
                'value' => $email_data
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $email_data)->where('variable', 'email_settings')->update('settings');
        }
    }



    public function update_rider_privacy_policy($data)
    {
        $data = escape_array($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'rider_privacy_policy'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'rider_privacy_policy',
                'value' => $data['privacy_policy_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $data['privacy_policy_input_description'])->where('variable', 'rider_privacy_policy')->update('settings');
        }
    }
    public function update_partner_privacy_policy($data)
    {
        $data = escape_array($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'partner_privacy_policy'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'partner_privacy_policy',
                'value' => $data['privacy_policy_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $data['privacy_policy_input_description'])->where('variable', 'partner_privacy_policy')->update('settings');
        }
    }

    public function update_rider_terms_n_condtions($data)
    {
        $data = escape_array($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'rider_terms_conditions'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'rider_terms_conditions',
                'value' => $data['terms_n_conditions_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $data['terms_n_conditions_input_description'])->where('variable', 'rider_terms_conditions')->update('settings');
        }
    }
    public function update_partner_terms_n_condtions($data)
    {
        $data = escape_array($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'partner_terms_conditions'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'partner_terms_conditions',
                'value' => $data['terms_n_conditions_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $data['terms_n_conditions_input_description'])->where('variable', 'partner_terms_conditions')->update('settings');
        }
    }
    public function update_admin_privacy_policy($data)
    {
        $data = escape_array($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'admin_privacy_policy'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'admin_privacy_policy',
                'value' => $data['privacy_policy_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $data['privacy_policy_input_description'])->where('variable', 'admin_privacy_policy')->update('settings');
        }
    }

    public function update_admin_terms_n_condtions($data)
    {
        $data = escape_array($data);
        $query = $this->db->get_where('settings', array(
            'variable' => 'admin_terms_conditions'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'admin_terms_conditions',
                'value' => $data['terms_n_conditions_input_description']
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $data['terms_n_conditions_input_description'])->where('variable', 'admin_terms_conditions')->update('settings');
        }
    }

    public function firebase_setting($post)
    {
        $post = escape_array($post);

        $system_data = json_encode($post);
        $query = $this->db->get_where('settings', array(
            'variable' => 'firebase_settings'
        ));
        $count = $query->num_rows();
        if ($count === 0) {
            $data = array(
                'variable' => 'firebase_settings',
                'value' => $system_data
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $system_data)->where('variable', 'firebase_settings')->update('settings');
        }
    }

    public function update_web_setting($post)
    {
        $post = escape_array($post);
        $post['app_download_section'] = (isset($post['app_download_section']) && !empty($post['app_download_section'])) ?: 0;
        $post['shipping_mode'] = (isset($post['shipping_mode']) && !empty($post['shipping_mode'])) ?: 0;
        $post['return_mode'] = (isset($post['return_mode']) && !empty($post['return_mode'])) ?: 0;
        $post['support_mode'] = (isset($post['support_mode']) && !empty($post['support_mode'])) ?: 0;
        $post['safety_security_mode'] = (isset($post['safety_security_mode']) && !empty($post['safety_security_mode'])) ?: 0;
        $main_image_name = (isset($post['logo']) && !empty($post['logo'])) ? $post['logo'] : "";
        $light_image_name = (isset($post['light_logo']) && !empty($post['light_logo'])) ? $post['light_logo'] : "";
        $favicon_image_name = (isset($post['favicon']) && !empty($post['favicon'])) ? $post['favicon'] : "";
        $system_data = json_encode($post);
        $query = $this->db->get_where('settings', array(
            'variable' => 'web_settings'
        ));
        $count = $query->num_rows();
        if ($main_image_name != NULL && !empty($main_image_name)) {
            $logo_res = $this->db->get_where('settings', array(
                'variable' => 'web_logo'
            ));
            $logo_count = $logo_res->num_rows();
            if ($logo_count == 0) {
                $this->db->insert('settings', ['value' => $main_image_name, 'variable' => 'web_logo']);
            } else {
                $this->db->set('value', $main_image_name)->where('variable', 'web_logo')->update('settings');
            }
        }
        if ($light_image_name != NULL && !empty($light_image_name)) {
            $logo_res = $this->db->get_where('settings', array(
                'variable' => 'light_logo'
            ));
            $logo_count = $logo_res->num_rows();
            if ($logo_count == 0) {
                $this->db->insert('settings', ['value' => $light_image_name, 'variable' => 'light_logo']);
            } else {
                $this->db->set('value', $light_image_name)->where('variable', 'light_logo')->update('settings');
            }
        }
        if ($favicon_image_name != NULL && !empty($favicon_image_name)) {
            $favicon_res = $this->db->get_where('settings', array(
                'variable' => 'web_favicon'
            ));
            $favicon_count = $favicon_res->num_rows();
            if ($favicon_count == 0) {
                $this->db->insert('settings', ['value' => $favicon_image_name, 'variable' => 'web_favicon']);
            } else {
                $this->db->set('value', $favicon_image_name)->where('variable', 'web_favicon')->update('settings');
            }
        }
        if ($count === 0) {
            $data = array(
                'variable' => 'web_settings',
                'value' => $system_data
            );
            $this->db->insert('settings', $data);
        } else {
            $this->db->set('value', $system_data)->where('variable', 'web_settings')->update('settings');
        }
    }
}

File: mobile/lib/data/model/settingModel.dart
class SettingModel {
  bool? error;
  int? allowModification;
  String? message;
  Data? data;

  SettingModel({this.error, this.allowModification, this.message, this.data});

  SettingModel.fromJson(Map<String, dynamic> json) {
    error = json['error'];
    allowModification = json['allow_modification'];
    message = json['message'];
    data = json['data'] != null ? Data.fromJson(json['data']) : null;
  }

  SettingModel copyWith({bool? error, int? allowModification, String? message, Data? data}) {
    return SettingModel(
      error: error ?? this.error,
      allowModification: allowModification ?? this.allowModification,
      message: message ?? this.message,
      data: data ?? this.data,
    );
  }
}

class Data {
  List<String>? logo;
  List<String>? privacyPolicy;
  List<String>? termsConditions;
  List<String>? fcmServerKey;
  List<String>? contactUs;
  List<String>? aboutUs;
  List<String>? currency;
  List<UserData>? userData;
  List<SystemSettings>? systemSettings;
  List<String>? tags;

  Data(
      {this.logo,
      this.privacyPolicy,
      this.termsConditions,
      this.fcmServerKey,
      this.contactUs,
      this.aboutUs,
      this.currency,
      this.userData,
      this.systemSettings,
      this.tags});

  Data.fromJson(Map<String, dynamic> json) {
    logo = json['logo'] == null ? List<String>.from([]) : (json['logo'] as List).map((e) => e.toString()).toList();
    privacyPolicy = json['privacy_policy'] == null ? List<String>.from([]) : (json['privacy_policy'] as List).map((e) => e.toString()).toList();
    termsConditions = json['terms_conditions'] == null ? List<String>.from([]) : (json['terms_conditions'] as List).map((e) => e.toString()).toList();
    fcmServerKey = json['fcm_server_key'] == null ? List<String>.from([]) : (json['fcm_server_key'] as List).map((e) => e.toString()).toList();
    contactUs = json['contact_us'] == null ? List<String>.from([]) : (json['contact_us'] as List).map((e) => e.toString()).toList();
    aboutUs = json['about_us'] == null ? List<String>.from([]) : (json['about_us'] as List).map((e) => e.toString()).toList();
    currency = json['currency'] == null ? List<String>.from([]) : (json['currency'] as List).map((e) => e.toString()).toList();

    if (json['user_data'] != null) {
      userData = <UserData>[];
      json['user_data'].forEach((v) {
        if (v.toString().isNotEmpty) {
          userData!.add(UserData.fromJson(v));
        }
      });
    }
    if (json['system_settings'] != null) {
      systemSettings = <SystemSettings>[];
      json['system_settings'].forEach((v) {
        systemSettings!.add(SystemSettings.fromJson(v));
      });
    }
    tags = json['tags'] == null ? List<String>.from([]) : (json['tags'] as List).map((e) => e.toString()).toList();
  }

  Data copyWith(
      {List<String>? logo,
      List<String>? privacyPolicy,
      List<String>? termsConditions,
      List<String>? fcmServerKey,
      List<String>? contactUs,
      List<String>? aboutUs,
      List<String>? currency,
      List<UserData>? userData,
      List<SystemSettings>? systemSettings,
      List<String>? tags}) {
    return Data(
        logo: logo ?? this.logo,
        privacyPolicy: privacyPolicy ?? this.privacyPolicy,
        termsConditions: termsConditions ?? this.termsConditions,
        fcmServerKey: fcmServerKey ?? this.fcmServerKey,
        contactUs: contactUs ?? this.contactUs,
        aboutUs: aboutUs ?? this.aboutUs,
        currency: currency ?? this.currency,
        userData: userData ?? this.userData,
        systemSettings: systemSettings ?? this.systemSettings,
        tags: tags ?? this.tags);
  }
}

class UserData {
  String? id;
  String? username;
  String? email;
  String? mobile;
  String? balance;
  String? dob;
  String? referralCode;
  String? friendsCode;
  String? cityName;
  String? area;
  String? landmark;
  String? pincode;
  String? cartTotalItems;

  UserData(
      {this.id,
      this.username,
      this.email,
      this.mobile,
      this.balance,
      this.dob,
      this.referralCode,
      this.friendsCode,
      this.cityName,
      this.area,
      this.landmark,
      this.pincode,
      this.cartTotalItems});

  UserData.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    username = json['username'];
    email = json['email'];
    mobile = json['mobile'];
    balance = json['balance'];
    dob = json['dob'];
    referralCode = json['referral_code'];
    friendsCode = json['friends_code'];
    cityName = json['city_name'];
    area = json['area'];
    landmark = json['landmark'];
    pincode = json['pincode'];
    cartTotalItems = json['cart_total_items'];
  }

  UserData copyWith(
      {String? id,
      String? username,
      String? email,
      String? mobile,
      String? balance,
      String? dob,
      String? referralCode,
      String? friendsCode,
      String? cityName,
      String? area,
      String? landmark,
      String? pincode,
      String? cartTotalItems}) {
    return UserData(
        id: id ?? this.id,
        username: username ?? this.username,
        email: email ?? this.email,
        mobile: mobile ?? this.mobile,
        balance: balance ?? this.balance,
        dob: dob ?? this.dob,
        referralCode: referralCode ?? this.referralCode,
        friendsCode: friendsCode ?? this.friendsCode,
        cityName: cityName ?? this.cityName,
        area: area ?? this.area,
        landmark: landmark ?? this.landmark,
        pincode: pincode ?? this.pincode,
        cartTotalItems: cartTotalItems ?? this.cartTotalItems);
  }
}

class SystemSettings {
  String? systemConfigurations;
  String? systemTimezoneGmt;
  String? systemConfigurationsId;
  String? appName;
  String? supportNumber;
  String? supportEmail;
  String? currentVersion;
  String? currentVersionIos;
  String? isVersionSystemOn;
  String? otpLogin;
  String? googleLogin;
  String? appleLogin;
  String? currency;
  String? systemTimezone;
  String? isReferEarnOn;
  String? isEmailSettingOn;
  String? minReferEarnOrderAmount;
  String? referEarnBonus;
  String? referEarnMethod;
  String? maxReferEarnAmount;
  String? referEarnBonusTimes;
  String? minimumCartAmt;
  String? lowStockLimit;
  String? maxItemsCart;
  String? isRiderOtpSettingOn;
  String? cartBtnOnList;
  String? expandProductImages;
  String? isAppMaintenanceModeOn;

  SystemSettings(
      {this.systemConfigurations,
      this.systemTimezoneGmt,
      this.systemConfigurationsId,
      this.appName,
      this.supportNumber,
      this.supportEmail,
      this.currentVersion,
      this.currentVersionIos,
      this.isVersionSystemOn,
      this.otpLogin,
      this.googleLogin,
      this.appleLogin,
      this.currency,
      this.systemTimezone,
      this.isReferEarnOn,
      this.isEmailSettingOn,
      this.minReferEarnOrderAmount,
      this.referEarnBonus,
      this.referEarnMethod,
      this.maxReferEarnAmount,
      this.referEarnBonusTimes,
      this.minimumCartAmt,
      this.lowStockLimit,
      this.maxItemsCart,
      this.isRiderOtpSettingOn,
      this.cartBtnOnList,
      this.expandProductImages,
      this.isAppMaintenanceModeOn});

  SystemSettings.fromJson(Map<String, dynamic> json) {
    systemConfigurations = json['system_configurations'];
    systemTimezoneGmt = json['system_timezone_gmt'];
    systemConfigurationsId = json['system_configurations_id'];
    appName = json['app_name'];
    supportNumber = json['support_number'];
    supportEmail = json['support_email'];
    currentVersion = json['current_version'];
    currentVersionIos = json['current_version_ios'];
    isVersionSystemOn = json['is_version_system_on'];
    otpLogin = json['otp_login'];
    googleLogin = json['google_login'];
    appleLogin = json['apple_login'];
    currency = json['currency'];
    systemTimezone = json['system_timezone'];
    isReferEarnOn = json['is_refer_earn_on'];
    isEmailSettingOn = json['is_email_setting_on'];
    minReferEarnOrderAmount = json['min_refer_earn_order_amount'];
    referEarnBonus = json['refer_earn_bonus'];
    referEarnMethod = json['refer_earn_method'];
    maxReferEarnAmount = json['max_refer_earn_amount'];
    referEarnBonusTimes = json['refer_earn_bonus_times'];
    minimumCartAmt = json['minimum_cart_amt'];
    lowStockLimit = json['low_stock_limit'];
    maxItemsCart = json['max_items_cart'];
    isRiderOtpSettingOn = json['is_rider_otp_setting_on'];
    cartBtnOnList = json['cart_btn_on_list'];
    expandProductImages = json['expand_product_images'];
    isAppMaintenanceModeOn = json['is_app_maintenance_mode_on'];
  }
}
